python/tests/compat/conftest.py

```diff
@@ -0,0 +1,142 @@
+import pytest
+import subprocess
+import os
+import venv
+import tempfile
+
+class ErrorInVenv(Exception):
+    """To signal errors occuring from within the venv"""
+
+def run_shell_commands(commands_list, cwd=None):
+    command = " && ".join(commands_list)
+    print("Executing command: ", command)
+    result = subprocess.run(command, cwd=cwd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    if result.returncode != 0 or True: # Fix
+        print("Command failed, logs:", result.stdout, result.stderr)
+    return result
+
+class Venv:
+    def __init__(self, venv_path, requirements_file):
+        self.path = venv_path
+        self.requirements_file = requirements_file
+
+    def __enter__(self):
+        self.init_venv()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.tear_down_venv()
+
+    def init_venv(self):
+        venv.create(self.path, with_pip=True, clear=True)
+        commands = [
+            f"bin/pip install -r {self.requirements_file}"
+        ]
+        run_shell_commands(commands, self.path)
+
+    def tear_down_venv(self):
+        commands = [
+            f"rm -rf {self.path}"
+        ]
+        run_shell_commands(commands)
+
+    def execute_python_file(self, python_path):
+        commands = [f"bin/python {python_path}"]
+        return run_shell_commands(commands, self.path)
+
+    def create_arctic(self, uri):
+        return VenvArctic(self, uri)
+
+
+class VenvArctic:
+    def __init__(self, venv, uri):
+        self.venv = venv
+        self.uri = uri
+        self.init_storage()
+
+    def execute(self, python_commands, dfs={}):
```

Comment:
A reference to the same empty dict default for `dfs` will be shared across invocations which might cause confusion or bugs. The normal way to do this in Python is,

```
def execute(dfs=None):
  if dfs is None:
    dfs = {}
  # stuff
```

Reply:
Thanks! Ugh python is weird.